lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/DriverTaskState.cs

```diff
@@ -0,0 +1,137 @@
+            { new TaskStateTransition(TaskState.TaskRunning, TaskEvent.FailedTaskCommunicationError), TaskState.TaskFailedByGroupCommunication },
+            { new TaskStateTransition(TaskState.TaskWaitingForClose, TaskEvent.ClosedTask), TaskState.TaskClosedByDriver },
+            { new TaskStateTransition(TaskState.TaskFailedSystemError, TaskEvent.FailedTaskEvaluatorError), TaskState.TaskFailedByEvaluatorFailure },
+            { new TaskStateTransition(TaskState.TaskFailedByGroupCommunication, TaskEvent.FailedTaskEvaluatorError), TaskState.TaskFailedByEvaluatorFailure }
+        });
+
+        private readonly static ISet<TaskState> FinalState = new HashSet<TaskState>()
+        {
+            TaskState.TaskFailedAppError,
+            TaskState.TaskFailedSystemError,
+            TaskState.TaskFailedByEvaluatorFailure,
+            TaskState.TaskFailedByGroupCommunication,
+            TaskState.TaskClosedByDriver,
+            TaskState.TaskCompleted
+        };
+
+        private volatile TaskState _currentState;
+        private readonly object _lock = new object();
+
+        /// <summary>
+        /// Create a new DriverTaskState with TaskNew as the task initial state
+        /// </summary>
+        internal DriverTaskState()
+        {
+            _currentState = TaskState.TaskNew;
+        }
+
+        /// <summary>
+        /// return the current task state
+        /// </summary>
+        internal TaskState CurrentState
+        {
+            get
+            {
+                return _currentState;
+            }
+        }
+
+        /// <summary>
+        /// Get next valid state based on the current state and event given without changing the current state
+        /// If there is no valid next state, ApplicationException will be thrown.
+        /// </summary>
+        /// <param name="taskEvent"></param>
+        /// <returns></returns>
+        internal TaskState GetNext(TaskEvent taskEvent)
+        {
+            TaskStateTransition transition = new TaskStateTransition(_currentState, taskEvent);
+            TaskState nextState;
+            if (!Transitions.TryGetValue(transition, out nextState))
+            {
+                throw new TaskStateTransitionException(_currentState, taskEvent);
+            }
+            return nextState;
+        }
+
+        /// <summary>
+        /// Move to the next state
+        /// If it is not able to move to the next valid state for a given event, ApplicationException will be thrown.
+        /// </summary>
+        /// <param name="taskEvent"></param>
+        /// <returns></returns>
+        internal TaskState MoveNext(TaskEvent taskEvent)
+        {
+            lock (_lock)
+            {
+                _currentState = GetNext(taskEvent);
+                return _currentState;
+            }
+        }
+
+        /// <summary>
+        /// Checks if the current state is a final state
```

Comment:
I have replied earlier for the same review comment above, in the code, most of the usage is that we have DriverTaskState object, if calling IsFinalState(CurrentState), we have to get current state then call it. Calling the first one is straightforward for a given DriverTaskState  object.